openvdb/openvdb/tree/NodeManager.h

```diff
@@ -530,13 +681,204 @@ class NodeManager
+{
+public:
+    DynamicNodeManagerLink() = default;
+
+    template<typename NodeOpT, typename RootT>
+    void foreachTopDown(const NodeOpT& op, RootT& root, bool threaded, size_t grainSize)
+    {
+        if (!op(root, /*index=*/0))         return;
+        if (!mList.initRootChildren(root))  return;
+        ForeachFilterOp<NodeOpT> filterOp(op, mList.nodeCount());
+        mList.foreachWithIndex(filterOp, threaded, grainSize);
+        mNext.foreachTopDownRecurse(filterOp, mList, threaded, grainSize);
+    }
+
+    template<typename FilterOpT, typename ParentT>
+    void foreachTopDownRecurse(const FilterOpT& filterOp, ParentT& parent, bool threaded, size_t grainSize)
+    {
+        if (!mList.initNodeChildren(parent, filterOp, !threaded))   return;
+        FilterOpT childFilterOp(filterOp.op(), mList.nodeCount());
+        mList.foreachWithIndex(childFilterOp, threaded, grainSize);
+    }
+
+    template<typename NodeOpT, typename RootT>
+    void reduceTopDown(NodeOpT& op, RootT& root, bool threaded, size_t grainSize)
+    {
+        if (!op(root, /*index=*/0))         return;
+        if (!mList.initRootChildren(root))  return;
+        ReduceFilterOp<NodeOpT> filterOp(op, mList.nodeCount());
+        mList.reduceWithIndex(filterOp, threaded, grainSize);
+        mNext.reduceTopDownRecurse(filterOp, mList, threaded, grainSize);
+    }
+
+    template<typename FilterOpT, typename ParentT>
+    void reduceTopDownRecurse(FilterOpT& filterOp, ParentT& parent, bool threaded, size_t grainSize)
+    {
+        if (!mList.initNodeChildren(parent, filterOp, !threaded))   return;
+        FilterOpT childFilterOp(filterOp.op(), mList.nodeCount());
+        mList.reduceWithIndex(childFilterOp, threaded, grainSize);
+    }
+
+protected:
+    NodeList<NodeT> mList;
+    DynamicNodeManagerLink<typename NodeT::ChildNodeType, LEVEL-1> mNext;
+};// DynamicNodeManagerLink class
+
+
+/// @private
+/// @brief Specialization that terminates the chain of cached tree nodes
+/// @note It is for internal use and should rarely be used directly.
+template<typename NodeT>
+class DynamicNodeManagerLink<NodeT, 0>
+{
+public:
+    DynamicNodeManagerLink() = default;
+
+    template<typename NodeFilterOp, typename ParentT>
+    void foreachTopDownRecurse(const NodeFilterOp& nodeFilterOp, ParentT& parent, bool threaded, size_t grainSize)
+    {
+        if (!mList.initNodeChildren(parent, nodeFilterOp, !threaded))   return;
+        mList.foreachWithIndex(nodeFilterOp.op(), threaded, grainSize);
+    }
+
+    template<typename NodeFilterOp, typename ParentT>
+    void reduceTopDownRecurse(NodeFilterOp& nodeFilterOp, ParentT& parent, bool threaded, size_t grainSize)
+    {
+        if (!mList.initNodeChildren(parent, nodeFilterOp, !threaded))   return;
+        mList.reduceWithIndex(nodeFilterOp.op(), threaded, grainSize);
+    }
+
+protected:
+    NodeList<NodeT> mList;
+};// DynamicNodeManagerLink class
+
+
+template<typename TreeOrLeafManagerT, Index _LEVELS>
```

Comment:
You mentioned you removed the ability to construct NodeManagers with LeafManagers - perhaps this should just be `TreeT`? Guess it will work either way, but you could then further guarantee the type with something like
`static_assert(std::is_base_of<TreeBase, TreeOrLeafManagerT>::value, "Type to DynamicNodeManager is not a valid tree type");`

Reply:
I removed the ability to construct NodeManagers from LeafManager **leaf arrays** - it is still possible to construct a NodeManager from a LeafManager as the call to `tree.root()` will map to `LeafManager::root()`. It's just that the root node will be used for the construction rather than the internal leaf array.